### PR TITLE
Feat/add bulker tests e2e

### PR DESCRIPTION
**Pull request by julien-devatom**
### Description of change

Closes: #60

### Pull-Request Checklist

- [ ] Code is up-to-date with the `main` branch
- [ ] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ ] The new commits follow conventions outlined in the [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/)

### CODE REVIEW

File: tests/e2e/bulker.test.ts [hunk 1]
```diff
@@ -0,0 +1,1021 @@
...
+
+    expect(await wAllowance()).to.be.equal(constants.Zero);
+    expect(await dAllowance()).to.be.equal(constants.Zero);
+    expect(await stAllowance()).to.be.equal(constants.Zero);
+    expect(await wstAllowance()).to.be.equal(constants.Zero);
+  });
+
+  const toAllow = [CONTRACT_ADDRESSES.bulker, CONTRACT_ADDRESSES.permit2];
+  toAllow.forEach((contractAddress) => {
+    const approval =
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:184]:
```suggestion
    const approvalType =
```

File: tests/e2e/bulker.test.ts [hunk 2]
```diff
@@ -0,0 +1,1021 @@
...
+          }
+          await bulker.executeBatch();
+
+          expect(maxWstethCapacity.limiter).to.equal(
+            MaxCapacityLimiter.walletBalance
+          );
+          const wstethBalanceLeft = await wsteth.balanceOf(morphoUser.address);
+          expect(wstethBalanceLeft).to.be.lessThan(
+            dust,
+            "wsteth balance left is less than 0.000001"
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:448]:
```suggestion
            "wsteth balance left is not less than 0.000001"
```

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:448]: Same everywhere

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:448]: it's mandatory to be able to compare the received value with the expected value

File: tests/e2e/bulker.test.ts [hunk 3]
```diff
@@ -0,0 +1,1021 @@
...
+          ).to.be.true;
+
+          expect(await wsteth.balanceOf(addresses.bulker)).to.be.equal(
+            constants.Zero,
+            "bulker wsteth balance is not 0"
+          );
+        });
+      });
+
+      it("Should supply collateral wstETH with some stETH to wrap and wstETH in wallet", async () => {
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:472]:
```suggestion
      it("Should supply partially wrapped stEth as collateral", async () => {
```

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:472]: same everywhere

File: tests/e2e/bulker.test.ts [hunk 4]
```diff
@@ -0,0 +1,1021 @@
...
+        await approve(contractAddress, async () => {
+          await wsteth.wrap(utils.parseEther("1"));
+          await morphoAdapter.refreshAll("latest");
+          await morphoAdapter.refetchData("latest");
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:476]: duplication with refreshAll

File: tests/e2e/bulker.test.ts [hunk 5]
```diff
@@ -0,0 +1,1021 @@
...
+            "steth balance left is not 0"
+          ).to.be.true;
+
+          const ma3Balance = await morphoAaveV3.collateralBalance(
+            wsteth.address,
+            morphoUser.address
+          );
+          expect(
+            approxEqual(ma3Balance, maxWstethCapacity.amount),
+            `ma3 balance (${ma3Balance}) is not equal to ${maxWstethCapacity.amount}`
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:516]:
```suggestion
            `expected ma3 balance to be ${maxWstethCapacity.amount}, received: (${ma3Balance}`
```

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:516]: Important to make clear what is expected and what is received

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:516]: Same everywhere

File: tests/e2e/bulker.test.ts [hunk 6]
```diff
@@ -0,0 +1,1021 @@
...
+        const wethBalance = await weth.balanceOf(morphoUser.address);
+        const finalAmount = amountToBorrow.add(initialWethBalance);
+        expect(wethBalance).to.equal(
+          finalAmount,
+          `weth balance (${wethBalance}) is not amountToBorrow + initialWethBalance (${finalAmount})`
+        );
+      });
+    });
+
+    it("should borrow ETH with a previous collateral position and unwrap", async () => {
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:574]:
```suggestion
    it("should borrow and unwrap ETH against an existing collateral position", async () => {
```

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:574]: same everywhere

File: tests/e2e/bulker.test.ts [hunk 7]
```diff
@@ -0,0 +1,1021 @@
...
+        const wethBalance = await weth.balanceOf(morphoUser.address);
+        expect(wethBalance).to.equal(
+          initialWethBalance,
+          "weth balance should not changed"
+        );
+
+        // Expect oldBalance < finalAmount < oldBalance + amount
+        // Trick to avoid gas fees with ETH
+        const finalAmount = await morphoUser.getBalance();
+        expect(oldBalance).to.be.lessThan(finalAmount);
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:621]:
```suggestion
        expect(initialBalance).to.be.lessThan(finalBalance);
```

Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:621]: or 
```suggestion
        expect(oldBalance).to.be.lessThan(newBalance);
```

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:621]: But please don't mix namings

File: tests/e2e/bulker.test.ts [hunk 8]
```diff
@@ -0,0 +1,1021 @@
...
+          );
+
+          const wethBalance = await weth.balanceOf(morphoUser.address);
+          // TODO I don't know how to solve this test, there's always remaining weth in user wallet
+          // expect(wethBalance).to.be.lessThan(
+          //   dust,
+          //   "weth balance should be unchanged (modulo dust)"
+          // );
+
+          // Expect oldBalance - 2 ETH < finalAmount < oldBalance - 1 ETH
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:858]: it's normal to have remaining weth, in order to repay max, you need to wrap a bit more of eth than the user's position so that the bulker contract have enough to repay the position + the interests accumulated during the blockj inclusion latency. Then the bulker contract sends back the excess of wrapped ETH

Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:858]: you have to check that the remaining amount is not greater than 0.01% of the borrow position ([that's the buffer we add](https://github.com/morpho-org/morpho-aave-v3-sdk/blob/12bff66972c0900a2b969333f1b5b2cd8cfe99eb/src/txHandler/Bulker.TxHandler.ts#L730-L753))

**Review by oumar-fall — CHANGES_REQUESTED** [review]
in summary:
- variable names shouldn't be shortened (`wAllowance` => `wethAllowance`)
- test suites names should be clearer
- please review [this pr](https://github.com/morpho-org/morpho-aave-v3-sdk/pull/70) and rebase
- test assertion error should always contain expected vs received value --> `expected [...] to be ${expectedValue}, received ${receivedValue}`
- no need to do `refreshAll` + `refetchData`
- the mechanism around the `approve` function is weird (why do we need to revoke the approval since we're reverting to a snapshot between every tests)? If really needed I would rather approve everything in the `beforeEach` function and revoke everything in the `afterEach`

File: tests/helpers/bn.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { BigNumber, BigNumberish } from "ethers";
+
+export const approxEqual = (value: BigNumberish, equality: BigNumberish) => {
+  const a = BigNumber.from(value);
+  const b = BigNumber.from(equality);
+  return a.sub(b).abs().lte(10);
```
[the remark this answers is not in the document]
Comment (oumar-fall) [reply tests/helpers/bn.ts:6]: i mean the `10` value

File: tests/e2e/bulker.test.ts [hunk 10]
```diff
@@ -0,0 +1,1021 @@
...
+  beforeEach(async () => {
+    snapshot = await hre.network.provider.send("evm_snapshot", []);
+    expect(await time.latestBlock()).to.be.equal(initialBlock);
+  });
+
+  afterEach(async () => {
+    bulker.reset();
+    await hre.network.provider.send("evm_revert", [snapshot]);
+    await morphoAdapter.refreshAll(initialBlock); // adapter reset
+    await morphoAdapter.refetchData(initialBlock); // adapter reset
```
[the remark this answers is not in the document]
Comment (ghivert) [reply tests/e2e/bulker.test.ts:113]: Weirdly enough, I had bugs when not using `refetchData`, I have no idea why

File: src/utils/signatures/withSigner.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import { Signer } from "ethers";
+import { splitSignature } from "ethers/lib/utils";
+
+import { MAX_UINT_160 } from "../../constants";
+import { BulkerSignature } from "../../txHandler/Bulker.TxHandler";
+
+export const fullfillBulkerSignWithSigner = async (
```
[the remark this answers is not in the document]
Comment (julien-devatom) [reply src/utils/signatures/withSigner.ts:7]: yep, deprecated (introduced before `Bulker.sign`)

File: src/txHandler/Bulker.TxHandler.ts [hunk 12]
```diff
@@ -819,6 +827,13 @@ export default class BulkerTxHandler
       signature: undefined,
       nonce: userData.nonce,
       transactionIndex: index,
+      getMessage: () =>
+        getManagerApprovalMessage(
+          userData.address,
+          addresses.bulker,
+          userData.nonce,
+          MAX_UINT_160
+        ),
```
[the remark this answers is not in the document]
Comment (julien-devatom) [reply src/txHandler/Bulker.TxHandler.ts:836]: mmmh idk, if we want to display some information about the signature using the typed object?

File: tests/e2e/bulker.test.ts [hunk 13]
```diff
@@ -0,0 +1,1021 @@
...
+  });
+
+  it("setup is well initialized", async () => {
```
[the remark this answers is not in the document]
Comment (julien-devatom) [reply tests/e2e/bulker.test.ts:116]: It's important imo, here you are validating the specs of the refetchData that you need for your tests. If this test breaks and by any chance the others are working, you are probably no longer testing the right things

File: tests/e2e/bulker.test.ts [hunk 14]
```diff
@@ -0,0 +1,1021 @@
...
+    const { address } = morphoUser;
+    const wAllowance = () => weth.allowance(address, CONTRACT_ADDRESSES.bulker);
```
[the remark this answers is not in the document]
Comment (julien-devatom) [reply tests/e2e/bulker.test.ts:118]: unless you are coding in go lol

File: src/txHandler/Bulker.TxHandler.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (oumar-fall) [reply src/txHandler/Bulker.TxHandler.ts:836]: It's still the same thing, do we need it now?

[the remark this answers is not in the document]
Comment (oumar-fall) [reply src/txHandler/Bulker.TxHandler.ts:836]: If we identify a need, then for sure but if not let's remove it for now (would take 10min to reimplement it when we need it)

File: tests/e2e/bulker.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:113]: Are you sure it wasn't before you put `latest` in argument?

File: tests/e2e/bulker.test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:116]: the refetchAll function should be testes with the adapter, not the bulker

File: src/txHandler/Bulker.TxHandler.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (oumar-fall) [reply src/txHandler/Bulker.TxHandler.ts:836]: another argument is that the day we need it, we might identify a better implementation than this one (more adapted to our real need)

[the remark this answers is not in the document]
Comment (julien-devatom) [reply src/txHandler/Bulker.TxHandler.ts:836]: Imo no big deal to keep or remove it

File: tests/e2e/bulker.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ghivert) [reply tests/e2e/bulker.test.ts:113]: My computer is too fast for rxjs. The solution is to add a delay 🤦🏻‍♂️

File: tests/e2e/bulker.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (ghivert) [reply tests/e2e/bulker.test.ts:118]: I have the habit to shortcut a long of things when working with OCaml when lifetime of the variable is short 😅

**Comment by ghivert** [comment]
Most of reviews have been applied.

On the `approve` function (renamed to `approveBulkerOrPermit2`), it is unfortunately required, because I'm not allowing the same contract on each run. Sometimes it's the bulker, sometimes it's Permit2. So we can't put it in `beforeEach`.

@oumar-fall On another note, I'm not sure the renaming of test suite really makes a difference, but I'm not disturbed by the change neither. I feel like it's a little bit micro-managing the codebase while the tests names are explicit and that we don't have precise conventions on this. I think it would be good to have conventions then, to avoid renaming everything everytime. (I'm not even 100% sure the renaming will suit you based on your comment, to be honest. 😳 I think it can quickly lead to anyone having its own naming convention by habit, and renaming everything many times, for no real improvements of the codebase.)

On the naming variables, I fully agree though, namings should be consistent, that was a good point.

File: src/txHandler/Bulker.TxHandler.ts [hunk 15]
```diff
@@ -52,22 +53,23 @@ type FullfillableSignature<Fullfilled extends boolean = boolean> =
     ? { deadline: BigNumber; signature: Signature }
     : undefined;
 
-export interface BulkerTransferSignature<Fullfilled extends boolean = boolean> {
+interface BaseBulkerSignature<Fullfilled extends boolean> {
```
Comment (oumar-fall) [inline src/txHandler/Bulker.TxHandler.ts:56]: Why did you remove the default value?

File: tests/helpers/bn.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import { BigNumber, BigNumberish } from "ethers";
+
+const APPROX_EQUAL_THRESHOLD = 10;
+
+export const approxEqual = (value: BigNumberish, equality: BigNumberish) => {
```
Comment (oumar-fall) [inline tests/helpers/bn.ts:5]: I would rather name the params `a` & `b` (`ethers-utils`'s convention) or `value` & `target`

Comment (oumar-fall) [reply tests/helpers/bn.ts:5]: not blocking though

File: tests/e2e/bulker.test.ts [hunk 17]
```diff
@@ -0,0 +1,1026 @@
...
+  const initialDaiBalance = utils.parseEther("5000");
+  const initialStEthBalance = utils.parseEther("5");
+  let bulker: BulkerTxHandler;
+
+  /** Approve tokens spending for Bulker.
+   * We unfortunately can't rely on beforeEach because we need to test two
+   * different cases which are mutually exclusive.
+   * We need to test with a Bulker approval and with a Permit2 approval.
+   * The idea of the function is to allow the Bulker to be easily tested with both.
+   */
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:60]: If both situations are mutually exclusive, it should be tested in two different test suite

File: tests/e2e/bulker.test.ts [hunk 18]
```diff
@@ -0,0 +1,1026 @@
...
+        await bulker.executeBatch();
+
+        expect(maxDaiCapacity.limiter).to.equal(
+          MaxCapacityLimiter.walletBalance
+        );
+
+        const userDaiBalance = await dai.balanceOf(morphoUser.address);
+        expect(userDaiBalance).to.be.equal(
+          constants.Zero,
+          `expected user dai balance is ${constants.Zero}, received ${userDaiBalance}`
```
Comment (oumar-fall) [inline tests/e2e/bulker.test.ts:220]:
```suggestion
          `expected user dai balance is ${constants.Zero.toString()}, received ${userDaiBalance.toString()}`
```

**Review by oumar-fall — COMMENTED** [review]
format big numbers in messages

File: tests/e2e/bulker.test.ts (hunk 17 above)
Comment (oumar-fall) [reply tests/e2e/bulker.test.ts:60]: Not 100% categoric on this point, must be discussed

File: tests/e2e/bulker.test.ts (hunk 18 above)
Comment (ghivert) [reply tests/e2e/bulker.test.ts:220]: It will call `.toString` on the BigNumbers if this is omitted.

File: src/txHandler/Bulker.TxHandler.ts (hunk 15 above)
Comment (ghivert) [reply src/txHandler/Bulker.TxHandler.ts:56]: It's an error

**Review by oumar-fall — DISMISSED** [review]
